diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index 5af55fc..73b5386 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -7670,6 +7670,11 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
             if (!(p = lock_user(VERIFY_WRITE, arg2, arg3, 0)))
                 return -TARGET_EFAULT;
             ret = get_errno(safe_read(arg1, p, arg3));
+            static bool last_read_empty = false;
+            if (ret == 0 && last_read_empty) {
+                exit_group(1);
+            }
+            last_read_empty = (ret == 0);
             if (ret >= 0 &&
                 fd_trans_host_to_target_data(arg1)) {
                 ret = fd_trans_host_to_target_data(arg1)(p, ret);
--
2.17.1
